### ISSUE TITLE
[labs/testing] Make WTR plugin work with on-the-fly ts compilation

### ISSUE THREAD

**Issue by augustjk** [lit/lit#3557]
### Should this be an RFC?

- [X] This is not a substantial change

### Which package is this a feature request for?

Other/unknown (please mention in description)

### Description

Currently the labs/testing WTR plugin and fixture functions only work when all test files are pre-compiled to JS before running the test.

Many projects (such as [shoelace](https://github.com/shoelace-style/shoelace)) use the [esbuild plugin for web test runner](https://modern-web.dev/docs/dev-server/plugins/esbuild/) for on-the-fly compilation of ts files. This is useful for not having to remember to run ts compilation before running tests, or for quickly iterating on tests and running it in watch mode.

Currently, the plugin does not work with the ssr testing plugin as dynamic imports expect .js file extensions and the esbuild plugin only works on browser requests, not dynamic imports done on the dev server.

### Alternatives and Workarounds

Tests for packages in Lit do not use the esbuild plugin, and instead pre-compile all TS files to JS before running tests. Coordination of this is handled with [wireit](https://github.com/google/wireit) so that running the test command automatically runs the build command if needed.

To support on-the-fly compilation, the lit ssr testing plugin's server code would need to detect and apply the same esbuild plugin settings for server-side dynamic imports (in addition to browser requests).

**Comment by justinfagnani** [comment]
I don't think that our testing package should have direct integration with esbuild or really any particular build system. Maybe we can add a config option for an import function, the default implementation of which just calls `import()`. Hopefully someone using esbuild can provide an implementation that calls esbuild first.

**Comment by WickyNilliams** [comment]
An extension point around import might help. Do you imagine that we'd still need to write the compiled file to disk, so that a module can be materialised? Pseudocode:

```
async function myCustomImport(path) {
  const compiled = await compileWithEsBuild(path)
  await writeToDisk(path, compiled)
  return import(path)
}
```

node does not yet have good support for creating in-memory es modules via its `vm` module https://nodejs.org/api/vm.html#class-vmmodule

I suppose in the ideal would be that the file can be imported from the WTR server itself via http. That way all the same esbuild/rollup plugins are applied. But unfortunately importing from http(s) urls is not generally available in node either: https://nodejs.org/api/esm.html#https-and-http-imports

**Comment by justinfagnani** [comment]
The nice thing about an import hook is that whatever facilities are available in the Node environment are available to use in the hook, even as things change.

I don't actually know the best way to integrate a WTR build setup with a otherwise plain Node import/fs operation (this is one of many reasons I personally prefer to to builds as their own non-integrated step with an npm script). WTR obviously has a way to get the bytes for a particular file internally, maybe that can be called to get bytes to write to disk.

**Comment by WickyNilliams** [comment]
yeah of course.

i guess it feels like something is amiss if the testing package is implemented as a WTR plugin, but it does not integrate with how many people use WTR. i was hoping this would be a quick win to add to our DS. i'll revisit again in future when node is better equipped for this.

**Comment by kyubisation** [comment]
I have created a PR as a base for this issue: #4730
It would be possible to extend the PR to allow an easy configuration for supporting .ts files.
